{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eecdebb2-830c-4996-9a55-70c9b3b6267e",
   "metadata": {},
   "source": [
    "# function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e636f768-287a-4934-a98b-121b98ca538e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGCCCAAAT\n"
     ]
    }
   ],
   "source": [
    "# code to find the reverse complement of a DNA sequence\n",
    "\n",
    "input_DNA = \"ATTTGGGCC\"\n",
    "\n",
    "rev_com_DNA = input_DNA.replace(\"A\", \"t\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"T\", \"a\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"C\", \"g\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"G\", \"c\")\n",
    "\n",
    "rev_com_DNA = rev_com_DNA.upper()[::-1]\n",
    "\n",
    "print(rev_com_DNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "255e086a-b000-41c9-895d-1aacb9eb4e4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGCCTAACCGCTAGCGTCATCAAAT\n"
     ]
    }
   ],
   "source": [
    "input_DNA = \"ATTTGATGACGCTAGCGGTTAGGCC\"\n",
    "\n",
    "rev_com_DNA = input_DNA.replace(\"A\", \"t\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"T\", \"a\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"C\", \"g\")\n",
    "rev_com_DNA = rev_com_DNA.replace(\"G\", \"c\")\n",
    "\n",
    "rev_com_DNA = rev_com_DNA.upper()[::-1]\n",
    "\n",
    "print(rev_com_DNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f66c344-8f13-4212-8c6d-27d5ee4d2a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we can convert this piece of code into a function\n",
    "# defining a function\n",
    "\n",
    "def do_reverse_complement(input_DNA):\n",
    "    rev_com_DNA = input_DNA.replace(\"A\", \"t\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"T\", \"a\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"C\", \"g\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"G\", \"c\")\n",
    "\n",
    "    rev_com_DNA = rev_com_DNA.upper()[::-1]\n",
    "    return rev_com_DNA    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0231c621-0e50-4e6f-b9fe-3e6106837396",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTAACATCCTTACACAT\n",
      "TAATTTCATCATCCCCTACATT\n"
     ]
    }
   ],
   "source": [
    "# function call \n",
    "input_DNA_1 = \"ATGTGTAAGGATGTTAG\"\n",
    "input_DNA_2 = \"AATGTAGGGGATGATGAAATTA\"\n",
    "\n",
    "rev_complement_DNA_1 = do_reverse_complement(input_DNA_1)\n",
    "rev_complement_DNA_2 = do_reverse_complement(input_DNA_2)\n",
    "\n",
    "print(rev_complement_DNA_1)\n",
    "print(rev_complement_DNA_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "77cc0a90-9b15-44bb-8432-44edc220b548",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function with no return value\n",
    "\n",
    "def do_reverse_complement_no_return(input_DNA):\n",
    "    rev_com_DNA = input_DNA.replace(\"A\", \"t\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"T\", \"a\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"C\", \"g\")\n",
    "    rev_com_DNA = rev_com_DNA.replace(\"G\", \"c\")\n",
    "\n",
    "    rev_com_DNA = rev_com_DNA.upper()[::-1]\n",
    "\n",
    "    print(rev_com_DNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bef0b0a9-d408-476d-8ad3-d9e8013d59e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTAACATCCTTACACAT\n",
      "TAATTTCATCATCCCCTACATT\n"
     ]
    }
   ],
   "source": [
    "# function call \n",
    "input_DNA_1 = \"ATGTGTAAGGATGTTAG\"\n",
    "input_DNA_2 = \"AATGTAGGGGATGATGAAATTA\"\n",
    "\n",
    "do_reverse_complement_no_return(input_DNA_1)\n",
    "do_reverse_complement_no_return(input_DNA_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "933ab177-e0b4-43ef-b2a3-86db72882687",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example of empty function \n",
    "\n",
    "def empty_function_demo():\n",
    "    pass\n",
    "\n",
    "empty_function_demo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b1e516d7-acf9-4289-b45c-e0c5e0d5a57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example of non-empty function without any arguments\n",
    "\n",
    "# we will create a function which generates random DNA sequence of lenght 10\n",
    "\n",
    "import random\n",
    "\n",
    "def generate_random_dna():\n",
    "    nucleotides = ['T', 'A', 'G', 'C']\n",
    "    dna_sequence = \"\".join(random.choice(nucleotides) for i in range(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f75a921c-aedc-4c4e-95c9-2839cd092f88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'G'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucleotides = ['T', 'A', 'G', 'C']\n",
    "random.choice(nucleotides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a6c2b697-6d16-4705-a903-10805e5e059d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "def generate_random_dna():\n",
    "    nucleotides = ['T', 'A', 'G', 'C']\n",
    "    dna_sequence = \"\".join(random.choice(nucleotides) for i in range(10))\n",
    "\n",
    "    return dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d76d492f-b4a2-4b53-9b05-0a406601d5a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACAGCGTGCG'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_random_dna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c6cad201-4839-4bfe-b985-caf00cb09d55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GATTGGTTGA'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_random_dna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73c51a5b-bca3-4198-8f49-3dabfce7c48f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# where i will pass two DNA sequence and we have to print their GC content upto 2 decimals and compare them \n",
    "\n",
    "def calculate_and_compare_GC_content(DNA_seq1, DNA_seq2, sig_dec):\n",
    "    comparison_result = ''\n",
    "\n",
    "    GC_content_DNA_1 = gc_fraction(DNA_seq1)\n",
    "    GC_content_DNA_1 = round(GC_content_DNA_1, sig_dec)\n",
    "\n",
    "    GC_content_DNA_2 = gc_fraction(DNA_seq2)\n",
    "    GC_content_DNA_2 = round(GC_content_DNA_2, sig_dec)\n",
    "\n",
    "    if GC_content_DNA_1 > GC_content_DNA_2:\n",
    "        comparison_result = 'greater'\n",
    "    \n",
    "    else:\n",
    "        comparison_result = 'less'\n",
    "\n",
    "    return GC_content_DNA_1, GC_content_DNA_2, comparison_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ef7c1d4e-308b-41d8-8fc8-339f4d3212e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.3     0.6     DNA1 content islessthan DNA2\n"
     ]
    }
   ],
   "source": [
    "DNA_seq1 = generate_random_dna() \n",
    "DNA_seq2 = generate_random_dna()\n",
    "sig_dec = 2\n",
    "\n",
    "GC_content_DNA_1, GC_content_DNA_2, comparison_message = calculate_and_compare_GC_content(DNA_seq1, DNA_seq2, sig_dec)\n",
    "\n",
    "print(f'{GC_content_DNA_1:<8}{GC_content_DNA_2:<8}{\"DNA1 content is\" + comparison_message + \"than DNA2\":<20}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "df9f1c38-142d-4317-a5ce-29ad4f6e1fb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_random_dna():\n",
    "    nucleotides = ['T', 'A', 'G', 'C']\n",
    "    dna_sequence = \"\".join(random.choice(nucleotides) for i in range(49))\n",
    "\n",
    "    return dna_sequence\n",
    "\n",
    "def calculate_and_compare_GC_content(DNA_seq1, DNA_seq2, sig_dec = 2):\n",
    "    comparison_result = ''\n",
    "\n",
    "    GC_content_DNA_1 = gc_fraction(DNA_seq1)\n",
    "    GC_content_DNA_1 = round(GC_content_DNA_1, sig_dec)\n",
    "\n",
    "    GC_content_DNA_2 = gc_fraction(DNA_seq2)\n",
    "    GC_content_DNA_2 = round(GC_content_DNA_2, sig_dec)\n",
    "\n",
    "    if GC_content_DNA_1 > GC_content_DNA_2:\n",
    "        comparison_result = 'greater'\n",
    "    \n",
    "    else:\n",
    "        comparison_result = 'less'\n",
    "\n",
    "    return GC_content_DNA_1, GC_content_DNA_2, comparison_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "85fee7fa-0847-4651-b63a-7714f8df6e92",
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_seq1 = generate_random_dna() \n",
    "DNA_seq2 = generate_random_dna()\n",
    "# case 1: where we didn't include the keyword argument\n",
    "GC_content_DNA_1, GC_content_DNA_2, comparison_message = calculate_and_compare_GC_content(DNA_seq1, DNA_seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "98e86a34-5867-470c-9130-b97a3a88c8a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.51    0.51    DNA1 content islessthan DNA2\n"
     ]
    }
   ],
   "source": [
    "print(f'{GC_content_DNA_1:<8}{GC_content_DNA_2:<8}{\"DNA1 content is\" + comparison_message + \"than DNA2\":<20}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb91cb17-59c3-44cd-aa2c-287a36af5501",
   "metadata": {},
   "source": [
    "# dynamic number of arguments in function definition"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3bae101-239a-4fa3-888b-e8082e66c1d5",
   "metadata": {},
   "source": [
    "1st set of experiment:\n",
    "   - gene expression: 2 times; averaged the value\n",
    "\n",
    "2nd set of experiment;\n",
    "   - gene expression: 3 times, averaged the value\n",
    "\n",
    "3th set of experiment;\n",
    "   - gene expression: 5 times, averaged the value\n",
    "\n",
    "contraint: we have to make a single function, which takes the value of gene expression received in each experiment.\n",
    "   - for e.g. \n",
    "   - if the experiment was repeated 2 times, function should be able to hand a scenario of average_gene_expression(a, b)\n",
    "   - if the experiment was repeated 3 times, the function should handle the scenario of average_gene_expression(a,b,c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2ee062a2-0943-47ae-a38b-c0ebd146e165",
   "metadata": {},
   "outputs": [],
   "source": [
    "def average_gene_expression(*args):\n",
    "    sum_expression = 0\n",
    "    count_experiment = 0\n",
    "    for expression_value in args:\n",
    "        # sum_expression = sum_expression + expression_value \n",
    "        sum_expression += expression_value\n",
    "        count_experiment += 1\n",
    "\n",
    "    average_expression = sum_expression/count_experiment\n",
    "\n",
    "    return round(average_expression, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5277de4e-8949-4d16-b407-cf9cbff59a12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130.0\n"
     ]
    }
   ],
   "source": [
    "# case 1 : we have two repeatition of the experiment \n",
    "\n",
    "average_value_expression = average_gene_expression(120, 140)\n",
    "print(average_value_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "11f6679a-924b-4888-91aa-b27662ab9c55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "128.33\n"
     ]
    }
   ],
   "source": [
    "# case 2 : we have three repeatition of the experiment\n",
    "\n",
    "average_value_expression = average_gene_expression(120, 140, 125)\n",
    "print(average_value_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "38d8ee67-6b1c-4e03-aa44-fd2a4391f1a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "122.2\n"
     ]
    }
   ],
   "source": [
    "# case 3 : we have ten repeatition of the experiment\n",
    "\n",
    "average_value_expression = average_gene_expression(120, 140, 125, 100, 122, 120, 90, 144, 125, 136)\n",
    "print(average_value_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "afa60be6-d97b-404e-9993-44d06fb9db95",
   "metadata": {},
   "outputs": [],
   "source": [
    "def average_gene_expression_keyword(sign_dec=2, **kwargs):\n",
    "    sum_expression = 0\n",
    "    count_experiment = 0\n",
    "    for id, expression_value in kwargs.items():\n",
    "        # sum_expression = sum_expression + expression_value \n",
    "        sum_expression += expression_value\n",
    "        count_experiment += 1\n",
    "\n",
    "    average_expression = sum_expression/count_experiment\n",
    "\n",
    "    return round(average_expression, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "20637994-a873-4492-819f-e53fefdbde23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "125.0"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2 experiments\n",
    "\n",
    "average_gene_expression_keyword(sign_dec=2, experiment=120, experiment2=130)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6566eab4-a346-4a81-87f7-2d4727467ad2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "140.8"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "average_gene_expression_keyword(sign_dec=2, experiment=120, experiment2=130, \n",
    "                                experiment3=150, experiment4=144, experiment5=160)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9bb4125-b2dd-4465-a486-07f285ee7577",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
